tools/celtest/test_runner.go

```diff
@@ -0,0 +1,698 @@
+		t.Run(test.Name, func(t *testing.T) {
+			err := tr.ExecuteTest(t, programs, test)
+			if err != nil {
+				t.Fatalf("error executing test: %v", err)
+			}
+		})
+	}
+}
+
+// TestSuiteParser provides a TestRunnerOption that sets the test suite file path and the test
+// suite parser based on the file format: YAML or Textproto.
+func TestSuiteParser(path string) TestRunnerOption {
+	return func(tr *TestRunner) (*TestRunner, error) {
+		if testSuitePath == "" {
+			return tr, nil
+		}
+		tr.TestSuiteFilePath = path
+		testSuiteFormat := compiler.InferFileFormat(testSuitePath)
+		switch testSuiteFormat {
+		case compiler.TextProto:
+			tr.TestSuiteParserTextproto = DefaultTestSuiteParserTextproto
+		case compiler.TextYAML:
+			tr.TestSuiteParserYAML = DefaultTestSuiteParserYAML
+		default:
+			return nil, fmt.Errorf("unsupported test suite file format: %v", testSuiteFormat)
+		}
+		return tr, nil
+	}
+}
+
+// TestRunner provides a structure to hold the different components required to execute tests for
+// a list of Input Expressions. The TestRunner can be configured with the following options:
+// - Compiler: The compiler used for parsing and compiling the input expressions.
+// - Input Expressions: The list of input expressions to be tested.
+// - Test Suite File Path: The path to the test suite file.
+// - File Descriptor Set Path: The path to the file descriptor set file.
+// - Test Suite Parser Textproto: A parser for a custom test suite file serialized in a Textproto
+// file. This option is required if the provided test suite is not a
+// cel.spec.expr.conformance.test.TestSuite message.
+// - Test Suite Parser YAML: A parser for a custom test suite file serialized in a YAML file.
+// This option is required if the provided test suite is not a TestSuite object.
+//
+// The TestRunner provides the following methods:
+// - Programs: Creates a list of CEL programs from the input expressions.
+// - Tests: Creates a list of tests from the test suite file.
+// - ExecuteTest: Executes a single
+type TestRunner struct {
+	compiler.Compiler
+	Expressions              []compiler.InputExpression
+	TestSuiteFilePath        string
+	FileDescriptorSetPath    string
+	TestSuiteParserTextproto TestSuiteParserTextproto
+	TestSuiteParserYAML      TestSuiteParserYAML
+}
+
+// Test represents a single test case to be executed. It encompasses the following:
+// - Name: The name of the test case.
+// - Input: The input to be used for evaluating the CEL expression.
+// - ResultMatcher: A function that takes in the result of evaluating the CEL expression and
+// returns a TestResult.
+type Test struct {
+	Name          string
+	Input         interpreter.Activation
+	ResultMatcher func(ref.Val, error) TestResult
+}
+
+// TestResult represents the result of a test case execution. It contains the validation result
+// along with the expected result and any errors encountered during the execution.
+type TestResult struct {
+	Success bool
```

Comment:
Does success mean the execution completed successfully or that the result matcher condition was satisfied? Consider adding a few comments on the fields to explain how they behave